File: 3.x.x/crates-utopia/cache/src/cache.rs
use std::collections::HashMap;
use std::sync::Arc;
use std::time::Instant;

use parking_lot::Mutex;
use utopia_telemetry::adapters::NoneAdapter;
use utopia_telemetry::{attrs, Adapter as TelemetryAdapter, Counter, Histogram};

use crate::adapter::Adapter;
use crate::error::CacheError;
use crate::value::{CacheValue, LoadResult, SaveResult};

const DURATION_NAME: &str = "cache.operation.duration";
const LOAD_TOTAL_NAME: &str = "cache.load.total";
const LOAD_TOTAL_DESC: &str = "Cache load operations broken down by hit/miss result.";

/// PHP `Utopia\Cache\Cache`.
pub struct Cache {
    adapter: Box<dyn Adapter>,
    /// PHP `$caseSensitive` (public).
    pub case_sensitive: bool,
    telemetry: Arc<dyn TelemetryAdapter>,
    operation_duration: Mutex<Option<Arc<dyn Histogram>>>,
    load_results: Mutex<Option<Arc<dyn Counter>>>,
}

impl std::fmt::Debug for Cache {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Cache")
            .field("case_sensitive", &self.case_sensitive)
            .finish_non_exhaustive()
    }
}

impl Cache {
    /// PHP `__construct(Adapter $adapter)`. Default telemetry is `None`.
    #[must_use]
    pub fn new(adapter: impl Adapter + 'static) -> Self {
        Self::from_boxed(Box::new(adapter))
    }

    #[must_use]
    pub fn from_boxed(adapter: Box<dyn Adapter>) -> Self {
        Self {
            adapter,
            case_sensitive: false,
            telemetry: Arc::new(NoneAdapter::new()),
            operation_duration: Mutex::new(None),
            load_results: Mutex::new(None),
        }
    }

    /// PHP `setTelemetry`. Instruments are created lazily on first use.
    pub fn set_telemetry(&mut self, telemetry: Arc<dyn TelemetryAdapter>) {
        self.telemetry = telemetry.clone();
        *self.operation_duration.lock() = None;
        *self.load_results.lock() = None;
        if let Some(inner) = self.adapter.as_telemetry_mut() {
            inner.set_telemetry(telemetry);
        }
    }

    fn get_operation_duration(&self) -> Arc<dyn Histogram> {
        let mut slot = self.operation_duration.lock();
        if let Some(h) = slot.as_ref() {
            return Arc::clone(h);
        }
        let mut advisory = HashMap::new();
        advisory.insert(
            "ExplicitBucketBoundaries".into(),
            "[0.001,0.005,0.01,0.025,0.05,0.1,0.25,0.5,1]".into(),
        );
        let hist = self
            .telemetry
            .create_histogram(DURATION_NAME, Some("s"), None, advisory);
        *slot = Some(Arc::clone(&hist));
        hist
    }

    fn get_load_results(&self) -> Arc<dyn Counter> {
        let mut slot = self.load_results.lock();
        if let Some(c) = slot.as_ref() {
            return Arc::clone(c);
        }
        let counter = self.telemetry.create_counter(
            LOAD_TOTAL_NAME,
            None,
            Some(LOAD_TOTAL_DESC),
            HashMap::new(),
        );
        *slot = Some(Arc::clone(&counter));
        counter
    }

    fn normalize<'a>(&self, value: &'a str) -> std::borrow::Cow<'a, str> {
        if self.case_sensitive {
            std::borrow::Cow::Borrowed(value)
        } else {
            std::borrow::Cow::Owned(value.to_lowercase())
        }
    }

    /// PHP `setCaseSensitivity`. Returns the assigned value.
    pub fn set_case_sensitivity(&mut self, value: bool) -> bool {
        self.case_sensitive = value;
        value
    }

    fn record_duration(&self, start: Instant, operation: &str, key: Option<&str>) {
        let duration = start.elapsed().as_secs_f64();
        let adapter_name = self.adapter.get_name(key);
        self.get_operation_duration().record(
            duration,
            &attrs(&[("operation", operation), ("adapter", adapter_name.as_str())]),
        );
    }

    /// PHP `load($key, $ttl, $hash = '')`.
    pub fn load(&self, key: &str, ttl: i64, hash: &str) -> Result<LoadResult, CacheError> {
        let key = self.normalize(key);
        let hash = self.normalize(hash);
        let start = Instant::now();
        let result = self.adapter.load(key.as_ref(), ttl, hash.as_ref())?;
        self.record_duration(start, "load", Some(key.as_ref()));
        let adapter_name = self.adapter.get_name(Some(key.as_ref()));
        let hit_or_miss = if result.is_miss() { "miss" } else { "hit" };
        self.get_load_results().add(
            1.0,
            &attrs(&[("adapter", adapter_name.as_str()), ("result", hit_or_miss)]),
        );
        Ok(result)
    }

    /// PHP `save($key, $data, $hash = '')`.
    pub fn save(
        &self,
        key: &str,
        data: impl Into<CacheValue>,
        hash: &str,
    ) -> Result<SaveResult, CacheError> {
        let key = self.normalize(key);
        let hash = self.normalize(hash);
        let data = data.into();
        let start = Instant::now();
        let result = self.adapter.save(key.as_ref(), &data, hash.as_ref());
        self.record_duration(start, "save", Some(key.as_ref()));
        result
    }

    /// PHP `getGeneration($key)`. Returns `"0"` when the adapter is not leasable.
    pub fn get_generation(&self, key: &str) -> Result<String, CacheError> {
        match self.adapter.as_leasable() {
            Some(leasable) => {
                let key = self.normalize(key);
                leasable.get_generation(key.as_ref())
            }
            None => Ok("0".into()),
        }
    }

    /// PHP `saveWithLease($key, $data, $hash, $generation)`.
    pub fn save_with_lease(
        &self,
        key: &str,
        data: impl Into<CacheValue>,
        hash: &str,
        generation: &str,
    ) -> Result<SaveResult, CacheError> {
        let key = self.normalize(key);
        let hash = self.normalize(hash);
        let data = data.into();
        let start = Instant::now();
        let result = if let Some(leasable) = self.adapter.as_leasable() {
            leasable.save_with_lease(key.as_ref(), &data, hash.as_ref(), generation)
        } else {
            self.adapter.save(key.as_ref(), &data, hash.as_ref())
        };
        self.record_duration(start, "saveWithLease", Some(key.as_ref()));
        result
    }

    /// PHP `touch($key, $hash = '')`.
    pub fn touch(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        let key = self.normalize(key);
        let hash = self.normalize(hash);
        let start = Instant::now();
        let result = self.adapter.touch(key.as_ref(), hash.as_ref());
        self.record_duration(start, "touch", Some(key.as_ref()));
        result
    }

    /// PHP `list($key)`.
    pub fn list(&self, key: &str) -> Result<Vec<String>, CacheError> {
        let key = self.normalize(key);
        let start = Instant::now();
        let result = self.adapter.list(key.as_ref());
        self.record_duration(start, "list", Some(key.as_ref()));
        result
    }

    /// PHP `purge($key, $hash = '')`.
    pub fn purge(&self, key: &str, hash: &str) -> Result<bool, CacheError> {
        let key = self.normalize(key);
        let hash = self.normalize(hash);
        let start = Instant::now();
        let result = self.adapter.purge(key.as_ref(), hash.as_ref());
        self.record_duration(start, "purge", Some(key.as_ref()));
        result
    }

    /// PHP `flush()`.
    pub fn flush(&self) -> Result<bool, CacheError> {
        let start = Instant::now();
        let result = self.adapter.flush();
        self.record_duration(start, "flush", None);
        result
    }

    /// PHP `ping()`.
    #[must_use]
    pub fn ping(&self) -> bool {
        self.adapter.ping()
    }

    /// PHP `getSize()`.
    pub fn get_size(&self) -> Result<i64, CacheError> {
        let start = Instant::now();
        let result = self.adapter.get_size();
        self.record_duration(start, "size", None);
        result
    }
}

File: packages/cache/src/Cache.php
<?php

namespace Utopia\Cache;

use Utopia\Cache\Feature\Batchable;
use Utopia\Cache\Feature\Leasable;
use Utopia\Telemetry\Adapter as Telemetry;
use Utopia\Telemetry\Adapter\None as NoTelemetry;
use Utopia\Telemetry\Counter;
use Utopia\Telemetry\Histogram;

class Cache
{
    /**
     * @var bool If cache keys are case-sensitive
     */
    public bool $caseSensitive = false;

    protected Telemetry $telemetry;

    protected ?Histogram $operationDuration = null;

    protected ?Counter $loadResults = null;

    /**
     * Set telemetry adapter. Instruments are created lazily on first use to
     * avoid emitting empty data point streams on every export interval.
     */
    public function setTelemetry(Telemetry $telemetry): void
    {
        $this->telemetry = $telemetry;
        $this->operationDuration = null;
        $this->loadResults = null;

        if ($this->adapter instanceof Feature\Telemetry) {
            $this->adapter->setTelemetry($telemetry);
        }
    }

    private function getOperationDuration(): Histogram
    {
        return $this->operationDuration ??= $this->telemetry->createHistogram(
            'cache.operation.duration',
            's',
            null,
            ['ExplicitBucketBoundaries' => [0.001, 0.005, 0.01, 0.025, 0.05, 0.1, 0.25, 0.5, 1]],
        );
    }

    private function getLoadResults(): Counter
    {
        return $this->loadResults ??= $this->telemetry->createCounter(
            'cache.load.total',
            null,
            'Cache load operations broken down by hit/miss result.',
        );
    }

    /**
     * Initialize with a no-op telemetry adapter by default.
     */
    public function __construct(private readonly Adapter $adapter)
    {
        $this->telemetry = new NoTelemetry();
    }

    /**
     * Toggle case sensitivity of keys inside cache
     *
     * @param  bool  $value if true, cache keys will be case-sensitive
     */
    public function setCaseSensitivity(bool $value): bool
    {
        return $this->caseSensitive = $value;
    }

    /**
     * Load cached data. return false in no valid cache.
     *
     * @param  int  $ttl time in seconds
     * @param  string  $hash optional
     */
    public function load(string $key, int $ttl, string $hash = ''): mixed
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        $hash = $this->caseSensitive ? $hash : strtolower($hash);

        $start = microtime(true);
        $result = $this->adapter->load($key, $ttl, $hash);
        $duration = microtime(true) - $start;
        $adapterName = $this->adapter->getName($key);
        $this->getOperationDuration()->record($duration, [
            'operation' => 'load',
            'adapter' => $adapterName,
        ]);
        $this->getLoadResults()->add(1, [
            'adapter' => $adapterName,
            'result' => $result === false ? 'miss' : 'hit',
        ]);

        return $result;
    }

    /**
     * Save data to cache. Returns data on success of false on failure.
     *
     * When $ttl > 0 the key is also given a key-level expiry (adapters that
     * support it arm it; others keep their timestamp TTL and ignore it). $ttl = 0
     * preserves the prior behaviour and leaves any expiry untouched.
     *
     * @param  string|array<int|string, mixed>  $data
     * @param  string  $hash optional
     * @param  int  $ttl time in seconds
     * @return bool|string|array<int|string, mixed>
     */
    public function save(string $key, mixed $data, string $hash = '', int $ttl = 0): bool|string|array
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        $hash = $this->caseSensitive ? $hash : strtolower($hash);
        $start = microtime(true);

        try {
            return $this->adapter->save($key, $data, $hash, $ttl);
        } finally {
            $duration = microtime(true) - $start;
            $this->getOperationDuration()->record($duration, [
                'operation' => 'save',
                'adapter' => $this->adapter->getName($key),
            ]);
        }
    }

    /**
     * Load several fields of $key in one round trip, as a field => value map with
     * missing/expired fields omitted. An empty $fields loads every field. Adapters
     * that do not store fields return an empty map.
     *
     * @param  int  $ttl time in seconds
     * @param  string[]  $fields
     * @return array<string, mixed>
     */
    public function loadMany(string $key, int $ttl, array $fields = []): array
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        if (! $this->caseSensitive) {
            $fields = array_map(strtolower(...), $fields);
        }

        $start = microtime(true);
        $result = $this->adapter instanceof Batchable
            ? $this->adapter->loadMany($key, $fields, $ttl)
            : [];
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'loadMany',
            'adapter' => $this->adapter->getName($key),
        ]);

        return $result;
    }

    /**
     * Write every field => value pair of $data in one round trip. When $ttl > 0
     * the key is given a key-level expiry. Returns $data on success, or false on
     * failure or for adapters that do not store fields.
     *
     * @param  array<string, mixed>  $data field => value
     * @param  int  $ttl time in seconds
     * @return array<string, mixed>|false
     */
    public function saveMany(string $key, array $data, int $ttl = 0): array|false
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        if (! $this->caseSensitive) {
            $data = array_change_key_case($data, CASE_LOWER);
        }

        $start = microtime(true);

        try {
            return $this->adapter instanceof Batchable
                ? $this->adapter->saveMany($key, $data, $ttl)
                : false;
        } finally {
            $duration = microtime(true) - $start;
            $this->getOperationDuration()->record($duration, [
                'operation' => 'saveMany',
                'adapter' => $this->adapter->getName($key),
            ]);
        }
    }

    /**
     * Current generation token for $key (advances on each purge). Returns '0'
     * when the underlying adapter does not support leasing.
     */
    public function getGeneration(string $key): string
    {
        if (! $this->adapter instanceof Leasable) {
            return '0';
        }

        $key = $this->caseSensitive ? $key : strtolower($key);

        return $this->adapter->getGeneration($key);
    }

    /**
     * Save $data only if the key's generation still equals $generation (i.e. no
     * purge happened since the caller captured it via getGeneration). Closes the
     * cache-aside read-after-write race. Falls back to an unconditional save when
     * the adapter does not support leasing.
     *
     * @param  string|array<int|string, mixed>  $data
     * @return bool|string|array<int|string, mixed>
     */
    public function saveWithLease(string $key, mixed $data, string $hash, string $generation): bool|string|array
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        $hash = $this->caseSensitive ? $hash : strtolower($hash);
        $start = microtime(true);

        try {
            if (! $this->adapter instanceof Leasable) {
                return $this->adapter->save($key, $data, $hash);
            }

            return $this->adapter->saveWithLease($key, $data, $hash, $generation);
        } finally {
            $duration = microtime(true) - $start;
            $this->getOperationDuration()->record($duration, [
                'operation' => 'saveWithLease',
                'adapter' => $this->adapter->getName($key),
            ]);
        }
    }

    /**
     * Refresh a cache entry timestamp without replacing its data.
     *
     * @param  string  $hash optional
     */
    public function touch(string $key, string $hash = ''): bool
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        $hash = $this->caseSensitive ? $hash : strtolower($hash);

        $start = microtime(true);
        $result = $this->adapter->touch($key, $hash);
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'touch',
            'adapter' => $this->adapter->getName($key),
        ]);

        return $result;
    }

    /**
     * Returns a list of keys.
     *
     * @return string[]
     */
    public function list(string $key): array
    {
        $key = $this->caseSensitive ? $key : strtolower($key);

        $start = microtime(true);
        $result = $this->adapter->list($key);
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'list',
            'adapter' => $this->adapter->getName($key),
        ]);

        return $result;
    }

    /**
     * Removes data from cache. Returns true on success of false on failure.
     *
     * @param  string  $hash optional
     */
    public function purge(string $key, string $hash = ''): bool
    {
        $key = $this->caseSensitive ? $key : strtolower($key);
        $hash = $this->caseSensitive ? $hash : strtolower($hash);

        $start = microtime(true);
        $result = $this->adapter->purge($key, $hash);
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'purge',
            'adapter' => $this->adapter->getName($key),
        ]);

        return $result;
    }

    /**
     * Removes all data from cache. Returns true on success of false on failure.
     */
    public function flush(): bool
    {
        $start = microtime(true);
        $result = $this->adapter->flush();
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'flush',
            'adapter' => $this->adapter->getName(),
        ]);

        return $result;
    }

    /**
     * Check Cache Connecitivity
     */
    public function ping(): bool
    {
        return $this->adapter->ping();
    }

    /**
     * Get db size.
     */
    public function getSize(): int
    {
        $start = microtime(true);
        $result = $this->adapter->getSize();
        $duration = microtime(true) - $start;
        $this->getOperationDuration()->record($duration, [
            'operation' => 'size',
            'adapter' => $this->adapter->getName(),
        ]);

        return $result;
    }
}
